### PR TITLE
Remove lowering extension hooks, replace with pass infrastructure.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This removes the "extension hook" for handling custom IR nodes in the native target lowering. The same behaviour can be achieved through use of the compiler pass infrastructure which is the declared extension interface for building compiler customisations.

The changes to the code base comprise:

* Creating a new pass to handle lowering of "parfor" nodes, this is based on the standard lowering pass. Both the new "parfor lowering" and the current "lowering" passes are now implemented as an extension of a "base" lowering pass. The "parfor lowering" pass is used as the lowerer if `parallel=True`.
* Removing the lowering "extension hook" as it's not needed.
* Removing the registration of the above "extension hook" for Parfor nodes in the CPU context as it is not needed.

### CODE REVIEW

File: numba/core/typed_passes.py [hunk 1]
```diff
@@ -434,6 +435,20 @@ def run_pass(self, state):
         return True
 
 
+@register_pass(mutates_CFG=True, analysis_only=False)
+class NativeLowering(BaseNativeLowering):
+
```
Comment (sklam) [inline numba/core/typed_passes.py:440]: empty line. maybe put a simple docstring here.

File: numba/core/typed_passes.py [hunk 2]
```diff
@@ -434,6 +435,20 @@ def run_pass(self, state):
...
+    _name = "native_lowering"
+    _lowering_class = lowering.Lower
+
+
+@register_pass(mutates_CFG=True, analysis_only=False)
+class NativeParforLowering(BaseNativeLowering):
+
```
Comment (sklam) [inline numba/core/typed_passes.py:447]: empty line. maybe put a simple docstring here.

File: numba/core/typed_passes.py [hunk 3]
```diff
@@ -352,10 +353,10 @@ def run_pass(self, state):
         return True
 
 
-@register_pass(mutates_CFG=True, analysis_only=False)
-class NativeLowering(LoweringPass):
+class BaseNativeLowering(LoweringPass):
 
-    _name = "native_lowering"
+    _name = None
+    _lowering_class = None
```
Comment (sklam) [inline numba/core/typed_passes.py:359]: Maybe making this abstractclassmethod so we lock down how it can be modified a little bit.

**Review by sklam — COMMENTED** [review]
The changes are unexpectedly simply. I only have a few minor comments:

**Comment by stuartarchibald** [comment]
> The changes are unexpectedly simply. I only have a few minor comments:

Thanks for the review, comments should be resolved in 444108f.

**Pull request merged**